### PR TITLE
add ProxyAll client config option to yaml and setting screen #2497

### PR DESCRIPTION
**Pull request by fffw**
To verify, manually add an 'proxyall' option under 'client' section in lantern.yaml.

### CODE REVIEW

**Comment by fffw** [comment]
Oh wait!  When I change proxyall option from true to false when Lantern is running, it triggers "send on closed channel" panic described in https://github.com/getlantern/lantern/issues/2500 . Good news is --- it's reproducible.

Changing this option from false to true works without causing panic.

**Comment by myleshorton** [comment]
> Oh wait! When I change proxyall option from true to false when Lantern is running, it triggers "send on closed channel" panic described in #2500 . Good news is --- it's reproducible.

Nice!

**Comment by myleshorton** [comment]
OK I'm on this one @fffw. I think this will be the primary new feature in beta 6 actually.

**Comment by myleshorton** [comment]
I believe I'm experiencing #2480 with this change as well -- I basically select to proxy all, and that setting never gets saved to disk, and Lantern doesn't proxy all. I suspect it's something intermittent and possibly a race condition on startup. I reproduced this on OSX.

**Comment by xiam** [comment]
Hello guys.

Any chance we could also enable the proxy-all feature from a command line option? It would really help us spotting and fixing some important issues.

Thanks!

**Comment by myleshorton** [comment]
Yeah I agree that's useful for debugging.

**Comment by myleshorton** [comment]
OK this generally looks good to me @fffw, although there is a bug with no longer unsetting the pac file on exit. We could just make this more explicit with the following patch to flashlight.go:

```
diff --git src/github.com/getlantern/flashlight/flashlight.go src/github.com/getlantern/flashlight/flashlight.go
index 0d8b1a0..8a2dd8e 100644
--- src/github.com/getlantern/flashlight/flashlight.go
+++ src/github.com/getlantern/flashlight/flashlight.go
@@ -8,6 +8,7 @@ import (
 	"os"
 	"runtime"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/getlantern/fronted"
@@ -43,7 +44,8 @@ var (
 	configUpdates = make(chan *config.Config)
 	exitCh        = make(chan error, 1)
 
-	showui = true
+	showui    = true
+	exitFuncs = make([]func(), 0)
 )
 
 func init() {
@@ -233,11 +235,18 @@ func runClientProxy(cfg *config.Config) {
 	watchDirectAddrs()
 
 	go func() {
-		defer pacOff()
+		addExitFunc(func() {
+			pacOff()
+		})
 		exit(client.ListenAndServe(pacOn))
 	}()
 }
 
+// addExitFunc adds a function to be called before the application exits.
+func addExitFunc(exitFunc func()) {
+	exitFuncs = append(exitFuncs, exitFunc)
+}
+
 func applyClientConfig(client *client.Client, cfg *config.Config) {
 	autoupdate.Configure(cfg)
 	logging.Configure(cfg, version, buildDate)
@@ -336,6 +345,10 @@ func useAllCores() {
 // exit tells the application to exit, optionally supplying an error that caused
 // the exit.
 func exit(err error) {
+	for _, ex := range exitFuncs {
+		log.Debugf("Calling exit func")
+		ex()
+	}
 	exitCh <- err
 }
 
```

**Comment by myleshorton** [comment]
Hey @fffw I realized this should actually use a channel instead of a slice - both simpler and thread safe!

File: src/github.com/getlantern/flashlight/flashlight.go [hunk 1]
```diff
@@ -98,7 +98,6 @@ func doMain() error {
 
 	// Schedule cleanup actions
 	defer logging.Close()
-	defer pacOff()
```
Comment (fffw) [inline src/github.com/getlantern/flashlight/flashlight.go]: It's me introduced the bug of not turning off pac by moving this line to below to near the call to `pacOn`. We can move it back here, or wrap the defer in a func, but I think add `exitFuncs` solely for this purpose is not worthy. How do you think @myleshorton ?

**Comment by fffw** [comment]
Oh let me double check if it actually works!

File: src/github.com/getlantern/flashlight/flashlight.go (hunk 1 above)
Comment (myleshorton) [inline src/github.com/getlantern/flashlight/flashlight.go]: That defer only worked when the method blocked, which the new one doesn't do. I think either way is fine though.

**Comment by fffw** [comment]
@myleshorton Used channel for exit functions as per your suggestion. Please take a look once you have time, thanks.

**Comment by myleshorton** [comment]
LGTM thanks @fffw!

**Pull request merged**